### ISSUE TITLE
Create a Querqy version for Solr 8.x

### ISSUE THREAD

**Issue by renekrie** [renekrie/querqy#50]

**Comment by renekrie** [comment]
Lucene 8 no longer permits negative scores (https://issues.apache.org/jira/browse/LUCENE-4100). We will need a way to cope with this for negative boost queries. Ideas:

- Can we find the max. score and then map scores between 0 and 1?
- Boost the negated query: x => DOWN(100): y becomes UP(100): not y
- Add 10000000 (a large number) to all scores.

This should also be an issue for the LTR Solr/Lucene plugins. How do they solve the problem?

**Comment by MighTguY** [comment]
Instead of giving and adding a boost to all the documents, can we create buckets using function queries in Solr and Elasticsearch?

**Comment by renekrie** [comment]
You mean something like:

~~~
x =>
  UP: <function query>
~~~

?

How would you express a bucket in a function query? - We would still have to make sure that the function query doesn't return a negative value.

**Comment by renekrie** [comment]
@martin-g I've created a new branch solr_8 from the current master and merged your PR #49 into it. I've fixed the tests in querqy-lucene, tests in query-solr are still not passing. As far as I can see, this is mostly due to forbidden negative scores in Lucene 8 (I'll dig into it and try to find a solution).

However, many thanks for your work! - I think your contributions already gets us beyond half the way down to Querqy for Solr/Lucene 8.

A few notes that might be handy also for other contributors in the future:

- In Querqy, we want to cover every Solr/Lucene version. So we first check if a Querqy version that was already created still works with the next Solr/Lucene version and then only create a new version if required. In this case, we try to cover Lucene/Solr 8.0 first and then check if we need a new version for 8.1
   
- Querqy has two main release artefacts that are released separately: querqy-core and querqy-for-lucene. querqy-core (search-engine-independent) normally doesn't need to be released if there is a new Solr/Lucene version. That's why the pom.xml in the root folder is not the parent of neither  querqy-core nor querqy-for-lucene. However, it allows for multi-module builds during development and continuous testing on Travis.

- Querqy models its own, flexible field boosting model, separate from (sub-)query boosting. The standard approach, that uses explicit field weights (à la f1^2 f2^0.5), is just one implementation of it. A different approach is implemented via querqy.lucene.rewrite.prms.PRMSFieldBoost. It tries to derive field weights per query term using a probabilistic approach based on index stats --> we need to leave this model in the TermQuery and Scorer implementation, otherwise tests will fail ;-), especially for the querqy.lucene.rewrite.prms package and for the DependentTermQueryBuilder

**Comment by martin-g** [comment]
Thank you for the new branch, @renekrie !
I will try to investigate the last failing tests in the coming days.

Actually I wanted to create an Elasticsearch plugin for Querqy. Since I'm using Elasticsearch 7.x I figured that I need to upgrade `querqy-lucene`. Lucene 8.1.0 was the latest in Maven at the moment.

In the meantime my client decided to use Solr instead of ES :-/

**Comment by renekrie** [comment]
Interesting - I also had a client that took the same decision two weeks ago. I assume it's not the same client... ;-)

I had started an attempt to create a Querqy plugin for ES. I ran into problems to find a place for caching the automaton that is used for rule input matching. For ES it might be more useful to create a standalone query rewriting service (based just on querqy-core) which returns a JSON in the ES Query DSL which can then be passed to ES. That would not enable all Querqy features (such as 'document frequency correction') - so, alternatively, the service might just return an intermediate query representation and the final transformation into a Lucene query is done by a minimal QueryBuilder plugin for ES (which doesn't need caching).

Let me know if that's something you would be interested in - I'm trying to find some contributors who would help create such an external service...

**Comment by renekrie** [comment]
Tests are passing now. Broken tests were mainly related to:

- LUCENE-4100 (no negative scores allowed), which is problematic for negative boost queries but also 'fake' document frequencies vs. idf calculation (negative idf values are no longer tolerated)
- `org.apache.lucene.search.CollectionStatistics` and `org.apache.lucene.search.TermStatistics` no longer allowing 'unset' values
- LegacyBM25Similarity's combining of boost and k1 + 1, so that 'boost' shows up with value boost * (k1 + 1) in Solr's debug output.

Though tests are passing, we still have an issue with negative boosts vs LUCENE-4100 (no negative scores allowed). The current, tentative approach is to add a  FunctionQuery in case of a negative boost:

```
 new FunctionScoreQuery(new QueryValueSource(boostQuery, Math.abs(negBoost))
(see querqy.lucene.QueryParsingController#transformBoostQueries)
```

This uses the absolute value of the negative boost factor as the default value of the FunctionScoreQuery. Documents that are not matching the negative boost query will thus be positively boosted by this value. The problem is that docs that match the boostQuery will get a higher boost the better 'boostQuery' matches. Maybe we could just wrap boostQuery into a constant score query as using the tf*idf-related score of the boostQuery as a negative boost dimension is questionable anyway.

**Comment by renekrie** [comment]
The solution for the neg boosting is to use a function score query:

```
// The score of negBoost will be the higher, the better a doc matches 
// negBoostQuery (e.g. via BM25) and equal 0 if a doc doesn't match this query:

negBoost = new QueryValueSource(negBoostQuery, 0f);

// the resulting boost FunctionQuery that we will combine 
// with the main query will then be:

float boost; // the absolute value of the negative boost factor (= the 100 in DOWN(100)) 
boostFunctionQuery = new ReciprocalFloatFunction(negBoost, boost, boost, 1f))

// which is equal to 
// v = boost / (boost * negBoost + 1)

// This results in v = boost for documents that don't match the negative boost query  
// - we are thus positively boosting documents that don't match the negative 
// boost query instead of giving a negative boost to documents that match it

// v will become the lower, the better a document scores for a the negative boost query

```

**Comment by renekrie** [comment]
The solution for negative boosts was changed again - see release notes for details: https://github.com/renekrie/querqy/wiki#15-june-2019---querqy-44lucene8000

**Comment by renekrie** [comment]
Released as version 4.4.lucene800.0
